arm9/source/utils/nandcmac.c

```diff
@@ -67,15 +67,72 @@ u32 SetupSlot0x30(char drv) {
+    u8 upper_counter[0x1];
+    u8 lower_counter[0x1];
+    u32 magic_check[0x4];
+    u32 slot_offset = 0;
+	
+    if (cmac_type == CMAC_AGBSAVE) return (slot_offset); // Does not apply for 'agbsave.bin'.
+	
+    // Look for the `.SAV` magic header at expected bottom offsets.
+    if (fvx_qread(path, magic_check, 0x400, 0x4, NULL) != FR_OK) return 1;
+    if (*magic_check == 0x5641532E) slot_offset = 0x400;
+
+    if (fvx_qread(path, magic_check, 0x2200, 0x4, NULL) != FR_OK) return 1;
+    if (*magic_check == 0x5641532E) slot_offset = 0x2200;
+
+    if (fvx_qread(path, magic_check, 0x8200, 0x4, NULL) != FR_OK) return 1;
+    if (*magic_check == 0x5641532E) slot_offset = 0x8200;
+
+    if (fvx_qread(path, magic_check, 0x10200, 0x4, NULL) != FR_OK) return 1;
+    if (*magic_check == 0x5641532E) slot_offset = 0x10200;
+
+    if (fvx_qread(path, magic_check, 0x20200, 0x4, NULL) != FR_OK) return 1;
+    if (*magic_check == 0x5641532E) slot_offset = 0x20200;
+	
+    // Compare top and bottom slots' counter values to determine which is newer.
+    if (fvx_qread(path, upper_counter, 0x34, 0x1, NULL) != FR_OK) return 1;
+    if (fvx_qread(path, lower_counter, slot_offset+0x034, 0x1, NULL) != FR_OK) return 1;
+
+    if ((*upper_counter == 0xFF) && (*lower_counter == 0x00)); // Scenario #1: First save is initialized. Bottom slot is newer.
+    else if ((*upper_counter == 0x00) && (*lower_counter == 0xFF)) slot_offset = 0; // Scenario #2: (ie, 0x100 > 0xFF). Top slot is newer.
+    else if (*upper_counter > *lower_counter) slot_offset = 0; // Scenario #3: Top slot is newer.
+
+    return (slot_offset);
+}
+
+u32 CheckAgbSaveHeader(const char* path) {
     AgbSaveHeader agbsave;
+    u32 magic_check[0x4];
+    u32 slot_offset = 0;
     UINT br;
-    
-    if ((fvx_qread(path, &agbsave, 0, 0x200, &br) != FR_OK) || (br != 0x200))
-        return 1;
-    
+	
+    // Look for the '.SAV' magic header at the top slot offset.
+    if (fvx_qread(path, magic_check, 0, 0x4, NULL) != FR_OK) return 1;
+
+    // Look for the `.SAV` magic header at expected bottom slot offsets if top is not found.
+    if (*magic_check != 0x5641532E) {
+        if (fvx_qread(path, magic_check, 0x400, 0x4, NULL) != FR_OK) return 1;
+        if (*magic_check == 0x5641532E) slot_offset = 0x400;
+
+        if (fvx_qread(path, magic_check, 0x2200, 0x4, NULL) != FR_OK) return 1;
+        if (*magic_check == 0x5641532E) slot_offset = 0x2200;
+
+        if (fvx_qread(path, magic_check, 0x8200, 0x4, NULL) != FR_OK) return 1;
+        if (*magic_check == 0x5641532E) slot_offset = 0x8200;
+
+        if (fvx_qread(path, magic_check, 0x10200, 0x4, NULL) != FR_OK) return 1;
+        if (*magic_check == 0x5641532E) slot_offset = 0x10200;
+
+        if (fvx_qread(path, magic_check, 0x20200, 0x4, NULL) != FR_OK) return 1;
+        if (*magic_check == 0x5641532E) slot_offset = 0x20200;
+    }
+	
+    if ((fvx_qread(path, &agbsave, slot_offset, 0x200, &br) != FR_OK) || (br != 0x200)) return 1;
```

Comment:
`return 1` isn't a great idea when you're already returning other integers - try to `return 0` on error, and handle that sanely whenever `CheckAgbSaveHeader` is called.

That goes for all the `return 1`s above, too.